packages/sdk38/src/lcdclient.spec.ts

```diff
@@ -0,0 +1,857 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { assert, sleep } from "@cosmjs/utils";
+
+import { rawSecp256k1PubkeyToAddress } from "./address";
+import { Coin } from "./coins";
+import { isPostTxFailure } from "./cosmosclient";
+import { makeSignBytes } from "./encoding";
+import { LcdApiArray, LcdClient, normalizeArray } from "./lcdclient";
+import { parseLogs } from "./logs";
+import { MsgSend } from "./msgs";
+import { makeCosmoshubPath, Secp256k1Pen } from "./pen";
+import { TxsResponse } from "./restclient";
+import { SigningCosmosClient } from "./signingcosmosclient";
+import cosmoshub from "./testdata/cosmoshub.json";
+import {
+  faucet,
+  makeRandomAddress,
+  makeSignedTx,
+  nonNegativeIntegerMatcher,
+  pendingWithoutWasmd,
+  tendermintIdMatcher,
+  wasmd,
+  wasmdEnabled,
+} from "./testutils.spec";
+import { StdFee } from "./types";
+
+/** Deployed as part of scripts/wasmd/init.sh */
+export const deployedErc20 = {
+  codeId: 1,
+  source: "https://crates.io/api/v1/crates/cw-erc20/0.5.1/download",
+  builder: "cosmwasm/rust-optimizer:0.8.0",
+  checksum: "3e97bf88bd960fee5e5959c77b972eb2927690bc10160792741b174f105ec0c5",
+  instances: [
+    "cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5", // HASH
+    "cosmos1hqrdl6wstt8qzshwc6mrumpjk9338k0lr4dqxd", // ISA
+    "cosmos18r5szma8hm93pvx6lwpjwyxruw27e0k5uw835c", // JADE
+  ],
+};
+
+describe("LcdClient", () => {
+  const defaultRecipientAddress = makeRandomAddress();
+
+  it("can be constructed", () => {
+    const client = new LcdClient(wasmd.endpoint);
+    expect(client).toBeTruthy();
+  });
+
+  describe("withModules", () => {
+    interface CodeInfo {
+      readonly id: number;
+      /** Bech32 account address */
+      readonly creator: string;
+      /** Hex-encoded sha256 hash of the code stored here */
+      readonly data_hash: string;
+      readonly source?: string;
+      readonly builder?: string;
+    }
+
+    type WasmResponse<T> = WasmSuccess<T> | WasmError;
+
+    interface WasmSuccess<T> {
+      readonly height: string;
+      readonly result: T;
+    }
+
+    interface WasmError {
+      readonly error: string;
+    }
+
+    function isWasmError<T>(resp: WasmResponse<T>): resp is WasmError {
+      return (resp as WasmError).error !== undefined;
+    }
+
+    function unwrapWasmResponse<T>(response: WasmResponse<T>): T {
+      if (isWasmError(response)) {
+        throw new Error(response.error);
+      }
+      return response.result;
+    }
+
+    interface WasmModule extends Record<string, () => any> {
+      listCodeInfo: () => Promise<readonly CodeInfo[]>;
+    }
+
+    it("works for no modules", async () => {
+      const client = LcdClient.withModules({ apiUrl: wasmd.endpoint });
+      expect(client).toBeTruthy();
+    });
+
+    it("works for one module", async () => {
+      pendingWithoutWasmd();
+      function wasmClientRegisterer(base: LcdClient): WasmModule {
```

Comment:
This is a good example, and also a demo of how to bind the `LcdClient` so it works when we copy the function to another `this`. I guess we should document to avoid classes in the `LcdModuleSetup`

Reply:
I wanted to support classes but type system disallows it because a class has no index signature. Modules need to conform to `export type LcdModule = Record<string, (...args: any[]) => any>;`, i.e. a map with string keys and function members.